File: backend/src/routes/analytics.ts
import { Router, Response } from 'express';
import { prisma } from '../prisma';
import { authenticate, requireRole, AuthRequest } from '../middleware/auth';

const router = Router();

// GET /api/analytics/admin - System-wide stats
router.get('/admin', authenticate, requireRole(['ADMIN']), async (req: AuthRequest, res: Response) => {
  try {
    const [totalStudents, totalRecruiters, totalJobs, totalApplications] = await Promise.all([
      prisma.studentProfile.count(),
      prisma.recruiterProfile.count(),
      prisma.job.count(),
      prisma.application.count(),
    ]);

    // Recruiter status breakdown
    const recruitersPending = await prisma.recruiterProfile.count({ where: { status: 'PENDING' } });
    const recruitersApproved = await prisma.recruiterProfile.count({ where: { status: 'APPROVED' } });

    // Job status breakdown
    const openJobs = await prisma.job.count({ where: { isOpen: true } });

    // Applications status breakdown
    const appBreakdown = await prisma.application.groupBy({
      by: ['status'],
      _count: {
        id: true,
      },
    });

    const statusCounts = {
      APPLIED: 0,
      SHORTLISTED: 0,
      REJECTED: 0,
      HIRED: 0,
    };

    appBreakdown.forEach((item) => {
      if (item.status in statusCounts) {
        statusCounts[item.status as keyof typeof statusCounts] = item._count.id;
      }
    });

    // Students placed (at least one HIRED application)
    const placedStudentsCount = await prisma.studentProfile.count({
      where: {
        applications: {
          some: {
            status: 'HIRED',
          },
        },
      },
    });

    const placementRate = totalStudents > 0 ? Math.round((placedStudentsCount / totalStudents) * 100) : 0;

    res.json({
      summary: {
        totalStudents,
        totalRecruiters,
        totalJobs,
        totalApplications,
        placementRate,
        placedStudents: placedStudentsCount,
      },
      recruiters: {
        pending: recruitersPending,
        approved: recruitersApproved,
      },
      jobs: {
        open: openJobs,
        closed: totalJobs - openJobs,
      },
      applications: statusCounts,
    });
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Server error fetching admin analytics' });
  }
});

// GET /api/analytics/recruiter - Recruiter dashboard stats
router.get('/recruiter', authenticate, requireRole(['RECRUITER']), async (req: AuthRequest, res: Response) => {
  try {
    const userId = req.user?.id;
    
    const recruiter = await prisma.recruiterProfile.findUnique({
      where: { userId },
    });

    if (!recruiter) {
      return res.status(404).json({ error: 'Recruiter profile not found' });
    }

    // Get jobs by this recruiter
    const recruiterJobs = await prisma.job.findMany({
      where: { recruiterProfileId: recruiter.id },
      include: {
        _count: {
          select: { applications: true },
        },
      },
    });

    const totalJobs = recruiterJobs.length;
    const openJobs = recruiterJobs.filter((j) => j.isOpen).length;

    // Total applications for recruiter's jobs
    const totalApplications = recruiterJobs.reduce((acc, job) => acc + job._count.applications, 0);

    // Group by status for applications to recruiter's jobs
    const appBreakdown = await prisma.application.groupBy({
      by: ['status'],
      where: {
        job: {
          recruiterProfileId: recruiter.id,
        },
      },
      _count: {
        id: true,
      },
    });

    const statusCounts = {
      APPLIED: 0,
      SHORTLISTED: 0,
      REJECTED: 0,
      HIRED: 0,
    };

    appBreakdown.forEach((item) => {
      if (item.status in statusCounts) {
        statusCounts[item.status as keyof typeof statusCounts] = item._count.id;
      }
    });

    // Applications count by job title for charts
    const jobsChartData = recruiterJobs.map((job) => ({
      title: job.title,
      applications: job._count.applications,
      isOpen: job.isOpen,
    }));

    res.json({
      summary: {
        totalJobs,
        openJobs,
        closedJobs: totalJobs - openJobs,
        totalApplications,
      },
      applications: statusCounts,
      jobsChartData,
    });
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Server error fetching recruiter analytics' });
  }
});

// GET /api/analytics/student - Student stats
router.get('/student', authenticate, requireRole(['STUDENT']), async (req: AuthRequest, res: Response) => {
  try {
    const userId = req.user?.id;

    const student = await prisma.studentProfile.findUnique({
      where: { userId },
    });

    if (!student) {
      return res.status(404).json({ error: 'Student profile not found' });
    }

    const myApps = await prisma.application.findMany({
      where: { studentProfileId: student.id },
    });

    const totalApplications = myApps.length;

    const statusCounts = {
      APPLIED: 0,
      SHORTLISTED: 0,
      REJECTED: 0,
      HIRED: 0,
    };

    myApps.forEach((app) => {
      if (app.status in statusCounts) {
        statusCounts[app.status as keyof typeof statusCounts]++;
      }
    });

    const totalOpenJobs = await prisma.job.count({
      where: { isOpen: true },
    });

    res.json({
      totalApplications,
      applications: statusCounts,
      totalOpenJobs,
    });
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Server error fetching student analytics' });
  }
});

export default router;

File: backend/routes/analytics.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from sqlalchemy import func
from database import get_db
from models import Application, Job, StudentProfile, RecruiterProfile
from auth import get_current_user

router = APIRouter(prefix="/api/analytics", tags=["analytics"])

@router.get("/student")
def student_analytics(db: Session = Depends(get_db), user: dict = Depends(get_current_user)):
    if user.get("role") != "STUDENT":
        raise HTTPException(status_code=403, detail="Forbidden")
    student = db.query(StudentProfile).filter(StudentProfile.userId == user["userId"]).first()
    if not student:
        raise HTTPException(status_code=404, detail="Student profile not found")
    apps = db.query(Application).filter(Application.studentProfileId == student.id).all()
    status_counts = {"APPLIED": 0, "SHORTLISTED": 0, "OFFERED": 0, "ACCEPTED": 0, "REJECTED": 0, "DECLINED": 0}
    for app in apps:
        s = app.status.value if hasattr(app.status, 'value') else str(app.status)
        if s in status_counts:
            status_counts[s] += 1
    total_open_jobs = db.query(func.count(Job.id)).filter(Job.isOpen == True).scalar()
    return {"totalApplications": len(apps), "applications": status_counts, "totalOpenJobs": total_open_jobs}

@router.get("/recruiter")
def recruiter_analytics(db: Session = Depends(get_db), user: dict = Depends(get_current_user)):
    if user.get("role") != "RECRUITER":
        raise HTTPException(status_code=403, detail="Forbidden")
    recruiter = db.query(RecruiterProfile).filter(RecruiterProfile.userId == user["userId"]).first()
    if not recruiter:
        raise HTTPException(status_code=404, detail="Recruiter profile not found")
    jobs = db.query(Job).filter(Job.recruiterProfileId == recruiter.id).all()
    total_jobs = len(jobs)
    open_jobs = sum(1 for j in jobs if j.isOpen)
    job_ids = [j.id for j in jobs]
    apps = db.query(Application).filter(Application.jobId.in_(job_ids)).all() if job_ids else []
    status_counts = {"APPLIED": 0, "SHORTLISTED": 0, "OFFERED": 0, "ACCEPTED": 0, "REJECTED": 0, "DECLINED": 0}
    for app in apps:
        s = app.status.value if hasattr(app.status, 'value') else str(app.status)
        if s in status_counts:
            status_counts[s] += 1
    app_count_by_job = {}
    for app in apps:
        app_count_by_job[app.jobId] = app_count_by_job.get(app.jobId, 0) + 1
    jobs_chart_data = [{"title": j.title, "applications": app_count_by_job.get(j.id, 0), "isOpen": j.isOpen} for j in jobs]
    return {"summary": {"totalJobs": total_jobs, "openJobs": open_jobs, "closedJobs": total_jobs - open_jobs, "totalApplications": len(apps)}, "applications": status_counts, "jobsChartData": jobs_chart_data}

@router.get("/admin")
def admin_analytics(db: Session = Depends(get_db), user: dict = Depends(get_current_user)):
    if user.get("role") != "ADMIN":
        raise HTTPException(status_code=403, detail="Forbidden")
    total_students = db.query(func.count(StudentProfile.id)).scalar()
    total_recruiters = db.query(func.count(RecruiterProfile.id)).scalar()
    total_jobs = db.query(func.count(Job.id)).scalar()
    total_apps = db.query(func.count(Application.id)).scalar()
    open_jobs = db.query(func.count(Job.id)).filter(Job.isOpen == True).scalar()
    apps = db.query(Application).all()
    status_counts = {"APPLIED": 0, "SHORTLISTED": 0, "OFFERED": 0, "ACCEPTED": 0, "REJECTED": 0, "DECLINED": 0}
    for app in apps:
        s = app.status.value if hasattr(app.status, 'value') else str(app.status)
        if s in status_counts:
            status_counts[s] += 1
    return {
        "summary": {"totalStudents": total_students, "totalRecruiters": total_recruiters, "totalJobs": total_jobs, "totalApplications": total_apps},
        "jobs": {"open": open_jobs, "closed": total_jobs - open_jobs},
        "applications": status_counts
    }
